feat: fisherman mode

spypsy opened the pull request:
Fixes [A-134](https://linear.app/aztec-labs/issue/A-134/fisherman-mode-for-the-node)  [A-147](https://linear.app/aztec-labs/issue/A-147/node-that-checks-every-attestation) [A-148](https://linear.app/aztec-labs/issue/A-148/node-that-checks-every-block-proposal) [A-149](https://linear.app/aztec-labs/issue/A-149/node-that-tries-to-build-block-on-every-slot) [A-150](https://linear.app/aztec-labs/issue/A-150/node-that-tries-to-invalidate-simulate-blocks) [A-256](https://linear.app/aztec-labs/issue/A-256/block-re-exeuction)

Add `--fisherman-mode [FISHERMAN_MODE]` option that will make the node:
- Always run validator
- re-execute all blocks & create attestation (that will NOT be broadcast)
- Validate all attestations against block proposal payload
- Attempts to build block for every slot
- Runs slashing logic & enqueues tx (never submits)

Also added alerting rules
- block re-execution failures
- block pre-validation / building failures

alexghr commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -996,19 +1021,24 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+    // In fisherman mode, we validate but don't penalize peers
+    const penalize = this.config.fishermanMode
+      ? (_severity: PeerErrorSeverity) => {}
+      : (severity: PeerErrorSeverity) => this.peerManager.penalizePeer(peerId, severity);
```
I wonder if this could be abused and have peers just constantly sending garbage our way and us not banning them?

spypsy replied:
hmm I guess if they specifically targeted the fisherman node, yes. The rest of the network will still penalize.
The rationale behind this was to not have any penalization locally on the fisherman node so we don't evict any peers and stop receiving data from them that may still be in the network

PhilWindle replied:
This is just the case where someone responds to a req/resp for a tx with something invalid. This doesn't apply to gossiping.

spypsy replied:
ah right yes this validation runs on ReqResp. Then we can go ahead and penalize in this validation

alexghr commented on yarn-project/validator-client/src/validator.ts:
```diff
@@ -325,17 +334,41 @@ export class ValidatorClient extends (EventEmitter as new () => WatcherEmitter)
+      return undefined;
+    }
+
+    return this.createBlockAttestationsFromProposal(proposal, attestors);
```
This ends up adding the attestations to this node's mempool. I'd say we should return early in fisherman mode (ie. we already re-executed and validated the output) rather than risking adding attestations from addresses not in the committee and risking breaking something else?

spypsy replied:
That was my original implementation. I ended up sending to the p2p layer and not broadcasting there, so we keep these attestations in the pool in case the data is useful for later lookup/analysis

spalladino replied:
I think I agree with Alex, I'd feel more comfortable if we don't push invalid stuff into our db. If we want to keep them for lookup, I'd just serialize them and log them, we can fish them from there later.

alexghr reviewed (commented):
IIUC we have to run the fisherman node with a set of keys for the validator (even thought the keys don't have to be registered with the rollup)?

spalladino commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -153,10 +159,13 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
   ) {
     super(telemetry, 'LibP2PService');
 
+    // Create child logger with fisherman prefix if in fisherman mode
+    this.logger = config.fishermanMode ? logger.createChild('[FISHERMAN]') : logger;
```
Curious: why?

spypsy replied:
just to make it easier to search for logs from components that are adjusted for fisherman mode

spalladino commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1038,7 +1065,10 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+      // In fisherman mode, we validate but don't penalize peers
+      if (!this.config.fishermanMode) {
+        this.peerManager.penalizePeer(peerId, severity);
+      }
```
Why not? I think we'll want our fisherman node to connect to healthy nodes, so it doesn't get an incorrect view of the p2p network if it gets stuck connected to faulty nodes that other nodes are dropping.

spypsy replied:
the idea was to not lose track of anything but I may be overthinking it. I think you're right that a few things can just work as in a normal node, including this

spalladino commented on yarn-project/sequencer-client/src/sequencer/metrics.ts:
```diff
@@ -36,6 +36,11 @@ export class SequencerMetrics {
+  private fishermanBlockBuildFailed: UpDownCounter;
+  private fishermanBlockBuildSuccess: UpDownCounter;
+  private fishermanPrecheckFailed: UpDownCounter;
+  private fishermanSlashingAttempts: UpDownCounter;
```
Why not make these regular metrics, as opposed to fisherman-only? It seems like valuable info to have for regular nodes as well.

spypsy replied:
tbh I just thought 'there must be a reason we don't have those as metrics already' which I really need to stop doing 😬

spalladino commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -385,15 +423,36 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
+          archive: block.archive.toString(),
+          txCount: block.body.txEffects.length,
+        });
+        // Store the fisherman block for later comparison (when actual proposal arrives)
```
I didn't understand this comment. What actual proposal?

spypsy replied:
bad AI comment for something that didn't need to be implemented, will remove

spalladino commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -358,15 +395,16 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
+    // Enqueues block invalidation (skip in fisherman mode)
+    if (invalidateBlock && !this.config.skipInvalidateBlockAsProposer && !this.config.fishermanMode) {
       publisher.enqueueInvalidateBlock(invalidateBlock);
     }
```
Why not enqueue anyway, as we are enqueuing the slashing actions, since the request is never actually sent?

spypsy replied:
I kept the slashing enqueueing because it also runs simulations in this step while others run the simulations separately. On your point of reducing `if (fishermanMode)` checks though, I agree we can just enqueue them all

spalladino commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -631,26 +696,39 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
+              proposerAddress ?? publisher.getSenderAddress(),
+            );
+
+      // In fisherman mode, skip enqueueing the block for publishing
```
Same here, why not enqueue anyway?

spalladino commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -311,8 +328,19 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
+      this.log.debug(`Using placeholder credentials for validation`, { coinbase, feeRecipient });
+    } else {
+      coinbase = this.validatorClient!.getCoinbaseForAttestor(attestorAddress);
+      feeRecipient = this.validatorClient!.getFeeRecipientForAttestor(attestorAddress);
```
I believe these two fallback to the attestor address itself if not set. Why not use those, to avoid another if?

spypsy replied:
hmm did not see that they fallback to something, will check

spalladino commented on yarn-project/p2p/src/msg_validators/attestation_validator/fisherman_attestation_validator.ts:
```diff
@@ -0,0 +1,91 @@
+import { AttestationValidator } from './attestation_validator.js';
+
+/**
+ * FishermanAttestationValidator extends the base AttestationValidator to add
+ * additional validation for fisherman nodes: verifying that attestations sign
+ * the same payload as the original proposal.
```
After Daniel's report, we'll have to add a check that an attestation matches a known proposal (see [A-268](https://linear.app/aztec-labs/issue/A-268/defend-against-block-proposal-flooding)). I wonder if we can just add that check there, without being too costly for a regular node (like instead of checking `hasBlockProposal`, also loading it and doing the same comparison as in here). 

Either way, I'm fine keeping this extended validator here until we define how we approach A268.

spalladino reviewed (commented):
Looks good, but in general it seems like many fisherman-specific things could instead be part of regular nodes?

- Fisherman should still penalize malicious p2p peers (IMO, happy to discuss)
- A regular node should emit block-building metrics same as a fisherman node
- We should be able enqueue proposals in the sequencers given we don't send them

All in all, we should be able to remove a bunch of `if fisherman` statements that make the code harder to follow.

spalladino commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1215,7 +1229,10 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+      // In fisherman mode, we validate but don't penalize peers
+      if (!this.config.fishermanMode) {
+        this.peerManager.penalizePeer(peerId, severity);
+      }
```
Not sure what we agreed, do we want to penalize anyway from fishermen? I vote yes, so we don't end up connected to malicious or stupid nodes that are not representative of the network, and other nodes would reject.

spypsy replied:
I think I made some changes on `v2` as I was testing it there that didn't transfer here 🤦 
thanks for spotting, will fix

spalladino commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1234,8 +1251,11 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+      if (!this.config.fishermanMode) {
+        this.logger.debug(`Penalizing peer ${peerId} for block proposal validation failure`);
+        this.peerManager.penalizePeer(peerId, severity);
+      }
```
Same

spalladino commented on yarn-project/sequencer-client/src/publisher/sequencer-publisher.ts:
```diff
@@ -914,29 +941,39 @@ export class SequencerPublisher {
+    if (this.config.fishermanMode) {
+      // In fisherman mode, we can't estimate blob gas because estimateGas doesn't support state overrides
+      // Use a fixed estimate.
+      blobEvaluationGas = BigInt(encodedData.blobs.length) * 21_000n;
```
I'm surprised 21k gas is enough, but hey, great!

spalladino commented on yarn-project/sequencer-client/src/publisher/sequencer-publisher.ts:
```diff
@@ -354,20 +376,25 @@ export class SequencerPublisher {
-        { address: MULTI_CALL_3_ADDRESS, balance },
-        ...(await this.rollupContract.makePendingBlockNumberOverride(opts?.forcePendingBlockNumber)),
-      ],
+      stateOverrides,
```
Heads up we're not sending the proposal from the multicall anymore here, not setting its balance. Not sure if it affects things though, where does the simulation run from if no `from` is set?

spypsy replied:
reverted

spalladino commented on yarn-project/sequencer-client/src/publisher/sequencer-publisher.ts:
```diff
@@ -354,20 +376,25 @@ export class SequencerPublisher {
-    const balance = await this.l1TxUtils.getSenderBalance();
-    this.log.debug(`Simulating validateHeader with balance: ${balance}`);
+    // In fisherman mode, we can't know where the proposer is publishing from
+    // so we simulate as the proposer but with sufficient balance
```
I don't understand why we need to simulate the proposal from a distinct address rather than coming from the multicall directly as we did before. I think I'm missing something in the overall simulation flow.

spalladino replied:
Ah, I guess that by default it is simulated as coming from one of the wallets in the viem client used by the sequencer publisher? And in most simulations we default to that, but this one was an exception for some reason where we set the `from` as the multicall?

spypsy replied:
actually I don't see a reason to remove `from: MULTI_CALL_3_ADDRESS` now that I look at it again. I think I was trying to simulate coming from the proposer to check their balance but then realized that's not the case anyway since they can be using a different publisher that we don't know about

spalladino commented on yarn-project/sequencer-client/src/sequencer/sequencer.ts:
```diff
@@ -289,28 +297,59 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
+        this.log.debug(`Set proposer address ${proposer} for simulation in fisherman mode`);
+      } else {
+        this.log.warn(
+          `Fisherman mode active but no specific proposer for slot ${slot} (anyone can propose). Simulations will use fisherman's own account which may have insufficient funds.`,
```
Why not use a state override to assign funds to it?

spypsy replied:
yeah we do now. Removing this log

spalladino commented on yarn-project/end-to-end/src/fixtures/utils.ts:
```diff
@@ -117,16 +117,26 @@ export const setupL1Contracts = async (
+      realVerifier: false,
+      ...args,
+    },
+    {
+      priorityFeeBumpPercentage: 0,
+      priorityFeeRetryBumpPercentage: 0,
+    },
```
Curious why this was needed?

spypsy replied:
I thought at first a test was failing bc of this. It was a different reason but still 
probably good to have so that we don't bid gas prices up in test environments during contract deployment

spypsy committed:
feat: fisherman mode

Fixes [A-134](https://linear.app/aztec-labs/issue/A-134/fisherman-mode-for-the-node)  [A-147](https://linear.app/aztec-labs/issue/A-147/node-that-checks-every-attestation) [A-148](https://linear.app/aztec-labs/issue/A-148/node-that-checks-every-block-proposal) [A-149](https://linear.app/aztec-labs/issue/A-149/node-that-tries-to-build-block-on-every-slot) [A-150](https://linear.app/aztec-labs/issue/A-150/node-that-tries-to-invalidate-simulate-blocks) [A-256](https://linear.app/aztec-labs/issue/A-256/block-re-exeuction)

Add `--fisherman-mode [FISHERMAN_MODE]` option that will make the node:
- Always run validator
- re-execute all blocks & create attestation (that will NOT be broadcast)
- Validate all attestations against block proposal payload
- Attempts to build block for every slot
- Runs slashing logic & enqueues tx (never submits)

Also added alerting rules
- block re-execution failures
- block pre-validation / building failures
